airflow/kubernetes/pod_usage_metrics_logger.py

```diff
@@ -0,0 +1,223 @@
+                          'Gi': 2 ** 20,
+                          'Ti': 2 ** 30,
+                          'Pi': 2 ** 40,
+                          'Ei': 2 ** 50,
+                          '': 1}
+    # Sometimes memory usage is not followed by unit if it is 0
+    # In these cases regex might return '' or None for unit.
+    valid_memory_usage_regex = re.compile(r'^([+]?[0-9]+?([.][0-9]+)?)(([KMGTPE][i])?)$')
+    # This regex checks for a positive float number with optional Ki, Mi, Gi, Ti, Pi, or Ei as unit
+
+    cpu_usage_units = {'m': 1,
+                       '': 1}  # Sometimes cpu usage is not followed by the unit when usage is 0
+    valid_cpu_usage_regex = re.compile(r'^([+]?[0-9]+?([.][0-9]+)?)([m]?)$')
+    # This regex check for a positive float number with optional millicpu(m) unit
+
+    def __init__(self,
+                 pod_launcher: object,  # (PodLauncher type) importing it might causes circular imports
+                 pod: V1Pod,
+                 resource_usage_fetch_interval: int,
+                 resource_usage_log_interval: int):
+        """
+        Creates the usage metrics logger.
+        :param pod_launcher: reference to the launcher of the metrics logger
+        :param pod: the pod we are logging the usage metrics
+        :param resource_usage_log_interval: logging interval in seconds
+        """
+        super().__init__()
+        self.pod_launcher = pod_launcher
+        self.pod = pod
+        self.resource_usage_fetch_interval = resource_usage_fetch_interval
+        self.resource_usage_log_interval = resource_usage_log_interval
+        self.max_cpu_usages = defaultdict(lambda: -1)  # type: dict
+        self.max_memory_usages = defaultdict(lambda: -1)  # type: dict
+
+    def log_pod_usage_metrics(self):
+        """
+        Periodically calls metrics api for the pod, logs current usage of the pod per container, and keeps
+        track of max usage seen for all containers.
+        :return:
+        """
+        if self.resource_usage_fetch_interval <= 0:
+            self.log.error('Parameter resource_usage_fetch_interval must be positive. Cancelling pod usage '
+                           'monitoring thread for pod: {0}.'.format(self.pod.metadata.name))
+            return
+        self.log.info('Pod usage monitoring thread started for pod: {0}'.format(self.pod.metadata.name))
+        resource_usage_api_url = "/apis/metrics.k8s.io/v1beta1/namespaces/{0}/pods/{1}"\
+                                 .format(self.pod.metadata.namespace,
+                                         self.pod.metadata.name)
+        cur_time = dt.now()
+        last_fetch_time = cur_time - timedelta(seconds=self.resource_usage_fetch_interval)
+        last_log_time = cur_time - timedelta(seconds=self.resource_usage_log_interval)
+        while self.pod_launcher.pod_is_running(self.pod):
+            if cur_time < last_fetch_time + timedelta(seconds=self.resource_usage_fetch_interval):
+                time.sleep(1)
```

Comment:
Why not just sleep for `self.resource_usage_fetch_interval`?

Reply:
We do not want to block main thread if the pod is not running anymore. Imagine a task with fetch interval set to 5min, then sleeping 5min while the pod is not running and blocking main thread is not ideal.